keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -230,80 +253,174 @@ scenario of wrong-single-stack pods...
+    and reverted the Service back to single-stack, it would get
+    re-reverted back to dual-stack by one of the running dual-stack
+    apiservers eventually.
+
+  - When the first dual-stack apiserver comes up, it would add its
+    primary IP to the `kubernetes` Endpoints and EndpointSlice, and
+    create a secondary EndpointSlice with its secondary IP.
+
+    Old apiservers would agree that the new apiserver's primary IP
+    belonged in the Endpoints and primary EndpointSlice, and would
+    ignore the secondary EndpointSlice, so we would get the right
+    result here as well.
+
+  - Eventually all of the apiservers would be running the new code,
+    and would all cooperatively maintain the dual-stack Service and
+    dual-stack EndpointSlices.
+
+The downgrade case will also (mostly) work, as long as the last of the
+dual-stack apiservers exits before the final downgraded single-stack
+apiserver starts (which should be the case if the apiservers are
+staying on the same hosts). In that case, the last downgraded
+apiserver will force the Service back to single-stack when it starts
+up. A stale secondary-IP-family EndpointSlice will be left behind
+after the downgrade, but nothing will be referring to it (since the
+Service will now be single-stack). (We can document these edge cases;
+see [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+below.)
 
-  - API server knows about `DualStackAPIServer`, the feature gate is
-    enabled, and the apiserver has a dual-stack
-    `--service-cluster-ip-range`:
+[kubernetes #101070]: https://github.com/kubernetes/kubernetes/issues/101070
 
-      - Advertises its own dual-stack endpoint IPs to etcd
+#### Upgrading from new single-stack apiservers to dual-stack apiservers
+
+Assuming we fix the "Service only reconciled at startup" and
+"secondary EndpointSlice not reconciled in single-stack cluster" bugs,
+there could potentially be flapping during an upgrade from
+single-stack apiserver to dual-stack apiserver, as the single-stack
+and dual-stack apiservers fight over the correct answer.
+
+The fix for that to update the rules so that apiservers can still
+reliably agree on what the Service and EndpointSlices should look
+like, even when they don't entirely agree on the current cluster
+configuration. Specifically, the rule we will use is that if _any_
+active apiserver is publishing single-stack EndpointSlices, then the
+cluster will have a single-stack apiserver. In detail:
+
+  - Periodically, the apiserver writes its own IP or IPs to etcd,
+    deletes any IPs with expired leases, and then gets the list of
+    remaining IPs. (This part is unchanged, other than the fact that
+    the apiserver may be writing dual-stack IPs to etcd.)
+
+  - Regardless of anything else, the apiserver always reconciles the
+    Endpoints (which are single-stack) and the primary-IP-family
+    EndpointSlice.
+
+  - If any active apiserver has written only single-stack Endpoints to
```

Comment:
Maybe worth mention: the primary IP family can not be omitted or else the errant apiserver will just be ignored (or something like that)